### PR TITLE
Feature Improvement: Custom Key Names for Structured Logging Support

### PR DESCRIPTION
**Pull request by SadiHassan**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
This is an extension of Structured logging feature which was merged in this PR #39277.
In this extension, user will be able to modify log key names. User will also be able to remove unwanted keys from log.
Consider this log below:
```{"date_time":"1650918987.180175","thread_name":"#1","thread_id":"254545","level":"Trace","query_id":"","logger_name":"BaseDaemon","message":"Received signal 2","source_file":"../base/daemon/BaseDaemon.cpp; virtual void SignalListener::run()","source_line":"192"}```. 

To enable JSON logging support, user need to uncomment the entire ```<formatting>``` tag below taken from `config.xml`.  

User will be able to modify key names by changing values under tag values inside <names> tag. For example, to change `date_time` to `MY_DATE_TIME`, user can do like:  
`<date_time>MY_DATE_TIME</date_time>`
 
User can stop unwanted log properties to appear in logs. To do so, they can simply comment out that property from this file.
For example, if you do not want your log to print query_id, you can comment out only `<query_id>` tag.However, if you comment out all the tags under <names>, the program will print default values for as below.  
   
        

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by evillique — COMMENTED** [review]
It seems like although the .xml file in the test has changed we do not check whether the labels that we assigned are in the logs or not.

**Comment by SadiHassan** [comment]
> It seems like although the .xml file in the test has changed we do not check whether the labels that we assigned are in the logs or not.

Yes, we don't. Last time Alexey asked to make a simple test case that will only check if a valid JSON log is the output or not. However, I am open to add more checking if you require.

File: src/Loggers/OwnJSONPatternFormatter.cpp [hunk 1]
```diff
@@ -8,93 +8,186 @@
...
-        writeJSONString("", wb, settings);
-    DB::writeChar(',', wb);
-
-    writeJSONString("source_line", wb, settings);
-    DB::writeChar(':', wb);
-    DB::writeChar('\"', wb);
-    DB::writeIntText(msg.getSourceLine(), wb);
-    DB::writeChar('\"', wb);
-
+    if (!this->date_time.empty())
```
[the remark this answers is not in the document]
Comment (SadiHassan) [reply src/Loggers/OwnJSONPatternFormatter.cpp:65]: Not necessary, will remove.

File: src/Loggers/OwnJSONPatternFormatter.h [hunk 2]
```diff
@@ -25,8 +25,20 @@ class Loggers;
...
 {
 public:
-    OwnJSONPatternFormatter();
+    OwnJSONPatternFormatter(Poco::Util::AbstractConfiguration & config_);
 
     void format(const Poco::Message & msg, std::string & text) override;
     void formatExtended(const DB::ExtendedLogMessage & msg_ext, std::string & text) const override;
+
+private:
+    Poco::Util::AbstractConfiguration & config;
```
[the remark this answers is not in the document]
Comment (SadiHassan) [reply src/Loggers/OwnJSONPatternFormatter.h:34]: I see, will modify it.

File: src/Loggers/OwnJSONPatternFormatter.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (SadiHassan) [reply src/Loggers/OwnJSONPatternFormatter.cpp]: Made changes and submitted.

File: src/Loggers/OwnJSONPatternFormatter.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (SadiHassan) [reply src/Loggers/OwnJSONPatternFormatter.h]: Made changes and submitted.

**Comment by SadiHassan** [comment]
Hi @evillique , I updated the integration test code to cover different scenarios of custom keys. Please kindly have a look and give me feedback for any changes required. Thanks!

**Comment by SadiHassan** [comment]
Hi @evillique , once you get a chance, can you please review my last updates and confirm?

**Review by evillique — APPROVED** [review]
Usually, tests with separate instances are made for testing communication between them, but the overhead does not seem to be large so it is probably ok.

File: src/Loggers/OwnJSONPatternFormatter.h [hunk 3]
```diff
@@ -25,8 +26,19 @@ class Loggers;
...
+
+private:
+    std::string date_time;
+    std::string thread_name;
+    std::string thread_id;
+    std::string level;
+    std::string query_id;
+    std::string logger_name;
+    std::string message;
+    std::string source_file_;
```
[the remark this answers is not in the document]
Comment (SadiHassan) [reply src/Loggers/OwnJSONPatternFormatter.h]: Changes submitted.

File: tests/integration/test_structured_logging_json/test.py [hunk 4]
```diff
@@ -23,10 +32,70 @@ def is_json(log_json):
...
+        ["cat", "/etc/clickhouse-server/config.d/config_all_keys_json.xml"]
+    )
+    config_some_keys = node_some_keys.exec_in_container(
+        ["cat", "/etc/clickhouse-server/config.d/config_some_keys_json.xml"]
+    )
+    config_no_keys = node_no_keys.exec_in_container(
+        ["cat", "/etc/clickhouse-server/config.d/config_no_keys_json.xml"]
+    )
+
+    assert (
```
[the remark this answers is not in the document]
Comment (SadiHassan) [reply tests/integration/test_structured_logging_json/test.py]: Changes submitted.

**Review by evillique — APPROVED** [review]

File: src/Loggers/Loggers.cpp [hunk 5]
```diff
@@ -99,8 +99,8 @@ void Loggers::buildLoggers(Poco::Util::AbstractConfiguration & config, Poco::Log
 
         Poco::AutoPtr<OwnPatternFormatter> pf;
 
-        if (config.getString("logger.formatting", "") == "json")
-            pf = new OwnJSONPatternFormatter;
+        if (config.getString("logger.formatting.type", "") == "json")
+           pf = new OwnJSONPatternFormatter(config);
```
Comment (evillique) [inline src/Loggers/Loggers.cpp:103]:
```suggestion
            pf = new OwnJSONPatternFormatter(config);
```

**Comment by evillique** [comment]
Failed check seem unrelated

**Pull request merged**